packages/server/src/games.ts

```diff
@@ -58,3 +58,99 @@ export function updateGameState(io: IOServer, store: Store, roomName: string, ga
     io.in(roomName).emit(GameEvent.CurrentState, updatedGameState);
   }
 }
+
+/**
+ * Calculate the immediate new game state, as a result of a player's action. This is not necessarily the final state
+ * but an intermediate one to help the following calculations.
+ */
+function calculatePlayerAction(oldState: GameState, playerAction: PlayerAction) {
+  const activePlayer = oldState.players.filter((playerState) => playerState.username === playerAction.playerName)[0];
+  // TODO: Later on we will need to add order of players, built into the model
+  const opponent = oldState.players.filter((playerState) => playerState.username !== playerAction.playerName)[0];
+  const hand = activePlayer.hand.filter((c) => cardKey(c) !== playerAction.card);
+  const table = [...oldState.table, fromCardKey(playerAction.card)];
+
+  const newState = {
+    ...oldState,
+    activePlayer: opponent.username,
+    players: [
+      opponent,
+      {
+        ...activePlayer,
+        hand,
+      },
+    ],
+    table,
+  };
+
+  return newState;
+}
+
+/**
+ * Calculate the new game state, as it should be at the end of a player's turn.
+ */
+function calculatePlayerTurn(oldState: GameState) {
+  const { players, deck, table, latestCaptured, ...rest } = oldState;
+
+  const isRoundFinshed = oldState.players.every((player) => player.hand.length === 0);
+  const isMatchFinished = isRoundFinshed && oldState.deck.length === 0;
+  let newState = <GameState>{};
+
+  // If match is finished (players and deck have no cards left), we add the remaining table
+  // cards to the player captured last and change the status to Ended.
+  if (isMatchFinished) {
+    players.forEach((player) => {
+      if (player.username === latestCaptured) {
+        player.captured.push(...table);
+      }
+    });
+
+    newState = {
+      ...rest,
+      status: GameStatus.Ended,
+      table,
+      players,
+      deck,
+      latestCaptured,
+    };
+  } else {
+    // If match has not finished but table is empty, this means a Scopa took place.
+    if (table.length === 0) {
+      players.forEach((player) => {
+        if (player.username === latestCaptured) {
+          player.scopa.push(last(player.captured) as Card);
+        }
+      });
+    }
+    if (isRoundFinshed) {
+      players.forEach((player) => {
+        player.hand = deck.splice(0, 3);
+      });
+    }
+    newState = {
+      ...rest,
+      table,
+      players,
+      deck,
+      latestCaptured,
+    };
+  }
+
+  return newState;
+}
+
+export function updateGameStateNew(io: IOServer, store: Store, roomName: string, playerAction: PlayerAction) {
+  const oldState = store.getRoomState(roomName);
+
+  if (!oldState) {
+    // TODO: Need to think about this, this should not be possible I think,
```

Comment:
true maybe we have to throw an error here anyway, not sure about errors though, how we should handle them and where

Reply:
Ah yeah, forgot about that... Will have another look but maybe I will leave for now.